sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,328 @@ def transform(self, X):
+                 weights="uniform", metric="masked_euclidean",
+                 row_max_missing=0.5, col_max_missing=0.8, copy=True):
+
+        self.missing_values = missing_values
+        self.n_neighbors = n_neighbors
+        self.weights = weights
+        self.metric = metric
+        self.row_max_missing = row_max_missing
+        self.col_max_missing = col_max_missing
+        self.copy = copy
+
+    def fit(self, X, y=None):
+        """Fit the imputer on X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape (n_samples, n_features)
+            Input data, where ``n_samples`` is the number of samples and
+            ``n_features`` is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        # Imports here to avoid circular import
+        from ..neighbors import NearestNeighbors
+        from ..neighbors.base import _check_weights
+
+        # Check parameters
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=np.float64,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        self.weights = _check_weights(self.weights)
+
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed.")
+
+        # Check if % missing in any column > col_max_missing
+        mask = _get_mask(X, self.missing_values)
+        if np.any(mask.sum(axis=0) > (X.shape[0] * self.col_max_missing)):
+            raise ValueError("Some column(s) have more than {}% missing values"
+                             .format(self.col_max_missing*100))
+        X_col_means = np.ma.array(X, mask=mask).mean(axis=0).data
+
+        # Check if % missing in any row > row_max_missing
+        bad_rows = mask.sum(axis=1) > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. These "
+                "rows are not included as donor neighbors."
+                    .format(self.row_max_missing * 100))
+
+            # Remove rows that have more than row_max_missing % missing
+            X = X[~bad_rows, :]
+
+        # Check if sufficient neighboring samples available
+        if X.shape[0] < self.n_neighbors:
+            raise ValueError("There are only %d samples, but n_neighbors=%d."
+                             % (X.shape[0], self.n_neighbors))
+
+        # Instantiate NN object, get column means, and store in statistics_
+        neigh = NearestNeighbors(n_neighbors=self.n_neighbors,
+                                 metric=self.metric,
+                                 metric_params={"missing_values":
+                                                self.missing_values})
+        self._fitted_neighbors = neigh.fit(X)
```

Comment:
Remind me why we wanted this to be private, not `neighbors_`. Did we want to leave open the option of not using it?

Reply:
To be honest, I also don't remember. I'll need to think about this a bit more.